### PR TITLE
Update algolia plugin

### LINKED ISSUE #1729: Update Algolia plugin
I'm having trouble getting the Algolia plugin to work with the latest search indexes provided by them (note that they migrated to a new infra recently).

Steps: 
1. Go to https://codesandbox.io/s/currying-leaf-qwo9k?file=/src/index.js  (this is a sandbox provided by Algolia devs to us)
2. Change 2021 to 2022 as indicated below
    image
3. Use the right-hand side to confirm the search is working

Now, I have trouble using the same config parameters with our algolia plugin. This is how my config looks like:
```json
    "algolia": {
      "apiKey": "a3fef7e2d9dea07c0fa2d8ed521057aa",
      "appId": "3DU4XTBX4T",
      "indexName": "nus-tic-cs",
      "algoliaOptions": { "hitsPerPage": 10 },
      "facetFilters": ["site:nus-tic2002-2022.github.io"],
      "debug": false
    }
```
Perhaps some updates needed to our Algolia plugin?

### PR DESCRIPTION
**Pull request by jonahtanjz**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [x] Others, please explain: Update Algolia plugin from v2 to v3

Fixes #1729 

**Overview of changes:**
Update Algolia from v2 to v3. Algolia now requires `appId` to be provided in order to use its APIs. V3 also requires that the search box be a `div` and no longer supports `input`.

User facing changes include adding new options and removing unsupported options:

**Added:** 
`appId` - compulsory field
`searchParameters` - optional

**Removed:**
`debug`
`algoliaOptions`

**Anything you'd like to highlight / discuss:**
V3 has changed its design drastically from V2. The search bar has a new design and clicking on the search bar will open a modal instead. The searches will be done on this modal and the results will be displayed on the modal instead of the dropdown that is used by Markbind's search. Since the new search bar and modal are auto generated by Algolia's package, it will be complicated to try to make the design similar to the current Markbind's search. Is it fine to stick to their native design? 

New search bar: 
image

New search modal instead of dropdown:
image

**Testing instructions:**
Add the following to `site.json` and serve the site.
```json
"plugins": [
    "algolia"
  ],
  "pluginsContext": {
    "algolia": {
      "apiKey": "599cec31baffa4868cae4e79f180729b",
      "appId": "R2IYF7ETH7",
      "indexName": "docsearch"
    }
  }
```

**Proposed commit message: (wrap lines at 72 characters)**
Update Algolia plugin

Algolia docsearch v2 is now a legacy version. Algolia also
requires new parameters for their APIs.

Let's update the version of Algolia docsearch to v3 and change
the plugin options accordingly.

---

**Checklist:** :ballot_box_with_check:

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [ ] No unrelated changes

### CODE REVIEW

**Comment by damithc** [comment]
Thanks for tackling this @jonahtanjz 

> V3 has changed its design drastically from V2. The search bar has a new design and clicking on the search bar will open a modal instead. The searches will be done on this modal and the results will be displayed on the modal instead of the dropdown that is used by Markbind's search. Since the new search bar and modal are auto generated by Algolia's package, it will be complicated to try to make the design similar to the current Markbind's search. Is it fine to stick to their native design?

I'm OK with it, if they haven't provided an easy way to use the dropdown for search results.

**Review by ryoarmanda — APPROVED** [review]
LGTM! 🎉

File: packages/core-web/src/index.js
Comment (ang-zeyu) [inline packages/core-web/src/index.js:60]: Just a quick look, would it be better to keep these strictly to the plugin itself? Since docsearch isn't really core functionality, but a plugin.

Some questions release wise:
- whether its possible to deliver a more minimal fix by adding the `apiKey` you mentioned onto `v3.1.1` or so (also so no immediate action needed on author end to migrate v2 -> v3)
- while we merge this bigger change into the main branch for `v4.0` (since v2 -> v3 docsearch is a breaking change)

**Comment by damithc** [comment]
Perhaps we can do a minor release after this is merged? I would like to start using this in the module website ASAP.

File: packages/core-web/src/index.js
Comment (jonahtanjz) [reply packages/core-web/src/index.js]: > Just a quick look, would it be better to keep these strictly to the plugin itself? Since docsearch isn't really core functionality, but a plugin.

I've shifted the plugin css code to within the plugin file. The rest of the plugin code related to `adjustHeaderClasses` is left there to minimise duplication as there are a couple of conditions that will need to adjust the plugin's css.

> * whether its possible to deliver a more minimal fix by adding the apiKey you mentioned onto v3.1.1 or so (also so no immediate action needed on author end to migrate v2 -> v3)
>
> * while we merge this bigger change into the main branch for v4.0 (since v2 -> v3 docsearch is a breaking change)

Yup agree with this. We should do the v2 -> v3 docsearch in our v4.0 and release the minimal fix in v3.1.1. I'll create a separate branch called `v3.1.x-hotfixes` (similar to the v2.7.x-hotfixes branch) and push the update there. Is this workflow fine?

Comment (ang-zeyu) [reply packages/core-web/src/index.js]: > Yup agree with this. We should do the v2 -> v3 docsearch in our v4.0 and release the minimal fix in v3.1.1. I'll create a separate branch called v3.1.x-hotfixes (similar to the v2.7.x-hotfixes branch) and push the update there. Is this workflow fine?

👍 

Alternatively there haven't been any breaking changes since the last release so far, so we could also merge the more minimal fix directly on the main branch. (I think either is fine)

Comment (ang-zeyu) [reply packages/core-web/src/index.js]: > I've shifted the plugin css code to within the plugin file. The rest of the plugin code related to adjustHeaderClasses is left there to minimise duplication as there are a couple of conditions that will need to adjust the plugin's css.


> V3 has changed its design drastically from V2. The search bar has a new design and clicking on the search bar will open a modal instead. The searches will be done on this modal and the results will be displayed on the modal instead of the dropdown that is used by Markbind's search. Since the new search bar and modal are auto generated by Algolia's package, it will be complicated to try to make the design similar to the current Markbind's search. Is it fine to stick to their native design?


Sorry, I completely missed the purpose of these changes earlier.

Are the dynamic styles added to tackle the above mentioned issue?

I'm favouring sticking with algolia's default design as much as possible if so, imo there isn't much benefit from pushing the dropdown / ui downward. Coupling our dynamic style detections over to an external element might break as well if algolia's styles / classes / dom structure changes.

If its an issue with correctness (e.g. the modal going out of the screen) I'll need a little more time to review this with a v3 algolia setup

Comment (jonahtanjz) [reply packages/core-web/src/index.js]: Mainly because the modal gets blocked by the navbar if the header is fixed.

image

I could try to increase the `z-index` of the modal to be higher. That may help to resolve this issue without pushing it down below the fixed header.

Comment (jonahtanjz) [reply packages/core-web/src/index.js]: Tested with z-index and it works as well. Will change to this implementation as it's much simpler.

image

File: packages/core-web/src/index.js
Comment (ang-zeyu) [inline packages/core-web/src/index.js:45]: Nice work @jonahtanjz, but do try to avoid the unnecessary changes here!

**Pull request merged**
